### PR TITLE
br: clean codes -- restore tables

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: ref #52877

Problem Summary:
we can clean some codes
### What changed and how does it work?
simplify the `split & scatter` and `download & ingest` stages
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/rtree/rtree.go [hunk 1]
```diff
@@ -85,8 +84,67 @@ func (rg *Range) Less(than btree.Item) bool {
 	return bytes.Compare(rg.StartKey, ta.StartKey) < 0
 }
 
+var _ btree.Item = &RangeStats{}
```
Comment (Leavrth) [inline br/pkg/rtree/rtree.go:87]: Range for backup, RangeStats for restore to be merged

File: br/pkg/task/restore.go [hunk 2]
```diff
@@ -1131,28 +1110,22 @@ func runRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConf
...
 	// Redirect to log if there is no log file to avoid unreadable output.
-	updateCh := g.StartProgress(
-		ctx,
-		cmdName,
-		progressLen,
-		!cfg.LogProgress)
+	updateCh := g.StartProgress(ctx, cmdName, progressLen, !cfg.LogProgress)
 	defer updateCh.Close()
-	sender, err := snapclient.NewTiKVSender(ctx, client, updateCh, cfg.PDConcurrency)
+	placementRuleManager, err := snapclient.NewPlacementRuleManager(ctx, mgr.GetPDClient(), mgr.GetPDHTTPClient(), mgr.GetTLSConfig(), cfg.Online)
```
Comment (Leavrth) [inline br/pkg/task/restore.go:1115]: There are many functions that set something before `split&scatter` and `download&ingest` and reset after that. We can unify them into one context manager.

File: br/pkg/restore/snap_client/batcher.go [hunk 3]
```diff
@@ -324,133 +114,29 @@ func (b *Batcher) filterOutRanges(checkpointSet map[string]struct{}, drained []r
 // |--|-------|
 // |t2|t3     |
 // as you can see, all restored ranges would be removed.
-func (b *Batcher) drainRanges() DrainResult {
+func drainRanges(
```
Comment (Leavrth) [inline br/pkg/restore/snap_client/batcher.go:117]: will be removed by the next PR

File: br/pkg/rtree/rtree.go [hunk 4]
```diff
@@ -85,8 +84,67 @@ func (rg *Range) Less(than btree.Item) bool {
...
+
+// RangeStats represents a restore merge result.
+type RangeStats struct {
+	*Range
```
Comment (YuJuncen) [inline br/pkg/rtree/rtree.go:91]: Perhaps no need to use pointer type?

File: br/pkg/rtree/rtree.go [hunk 5]
```diff
@@ -85,8 +84,67 @@ func (rg *Range) Less(than btree.Item) bool {
...
+
+// Less impls btree.Item.
+func (rg *RangeStats) Less(than btree.Item) bool {
+	// rg.StartKey < than.StartKey
+	ta := than.(*RangeStats)
+	return bytes.Compare(rg.StartKey, ta.StartKey) < 0
+}
+
+type RangeStatsTree struct {
+	*btree.BTree
```
Comment (YuJuncen) [inline br/pkg/rtree/rtree.go:104]: Perhaps use generic version of btree?

File: br/pkg/task/restore_raw.go [hunk 6]
```diff
@@ -145,7 +146,7 @@ func RunRestoreRaw(c context.Context, g glue.Glue, cmdName string, cfg *RestoreR
 		!cfg.LogProgress)
 
 	// RawKV restore does not need to rewrite keys.
-	err = client.SplitRanges(ctx, ranges, updateCh, true)
+	err = client.SplitRanges(ctx, getEndKeys(ranges), updateCh, true)
```
Comment (YuJuncen) [inline br/pkg/task/restore_raw.go:149]: Perhaps rename it to `SplitPoints` given the argument isn't `[]Range` anymore.

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/pkg/restore/utils/merge.go [hunk 7]
```diff
@@ -48,6 +51,14 @@ func MergeAndRewriteFileRanges(
 	for _, file := range files {
 		filesMap[string(file.StartKey)] = append(filesMap[string(file.StartKey)], file)
 
+		// Assert that it has the same end key.
+		if !bytes.Equal(filesMap[string(file.StartKey)][0].EndKey, file.EndKey) {
```
Comment (3pointer) [inline br/pkg/restore/utils/merge.go:55]: why need this check? for some scenarios like compacted ssts may not satisfied.

File: br/pkg/rtree/rtree.go [hunk 8]
```diff
@@ -85,8 +84,64 @@ func (rg *Range) Less(than btree.Item) bool {
 	return bytes.Compare(rg.StartKey, ta.StartKey) < 0
 }
 
+// RangeStats represents a restore merge result.
+type RangeStats struct {
+	Range
+	Size  uint64
```
Comment (3pointer) [inline br/pkg/rtree/rtree.go:90]: Size and Count seems never used...

**Review by 3pointer — APPROVED** [review]

File: br/pkg/rtree/rtree.go (hunk 8 above)
Comment (Leavrth) [reply br/pkg/rtree/rtree.go:90]: Used in the next PR

File: br/pkg/restore/utils/merge.go (hunk 7 above)
Comment (Leavrth) [reply br/pkg/restore/utils/merge.go:55]: To assert range's start-key and end-key is the same as that of any file of the range.

**Pull request merged**